Expose `conn` validation for reuse in custom plugs

soundmonster opened the pull request:
The validation plug is quite handy, but only works best at the top-level
router. This commit allows piping the whole API scope (e.g. /api)
throught the validator plug, but allowing Swagger UI to be hosted
somewhere under the scope's path (e.g. /api/v1/docs) without validation
errors.

mbuhot commented on lib/phoenix_swagger/plug/validate_plug.ex:
```diff
@@ -10,19 +10,31 @@ defmodule PhoenixSwagger.Plug.Validate do
   def call(conn, opts) do
-    validation_failed_status = Keyword.get(opts, :validation_failed_status, 400)
+    regex_result = opts
+    |> Keyword.get(:exclude_paths, [~r/$^/]) # the default regex matches nothing
```
`List.wrap()` will convert a `nil` into an empty list, so no need for the default regex.

mbuhot commented on lib/phoenix_swagger/plug/validate_plug.ex:
```diff
@@ -10,19 +10,31 @@ defmodule PhoenixSwagger.Plug.Validate do
+  @doc """
+  Performs unconditional validation of a request. Will halt and send error on failed validation.
+  """
+  def handle_validation(conn, opts) do
```
This feels like a private function

0xAX commented on lib/phoenix_swagger/plug/validate_plug.ex:
```diff
@@ -32,6 +44,21 @@ defmodule PhoenixSwagger.Plug.Validate do
+  * `{:error, message, path}` if the request was mapped but failed validation
+  """
+  def validate(conn) do
+    result =
```
this `result` variable is unused

0xAX commented:
Otherwise looks good to me

mbuhot commented:
> but allowing Swagger UI to be hosted
somewhere under the scope's path (e.g. /api/v1/docs) without validation
errors.

I believe this already works, eg see the example project:

```elixir
  pipeline :api do
    plug :accepts, ["json"]
    plug Validate, validation_failed_status: 422
  end

  scope "/api", SimpleWeb do
    pipe_through :api
    resources "/users", UserController, except: [:new, :edit]
  end

  scope "/api/swagger" do
    get "/swagger.json", SimpleWeb.SwaggerController, :show
    forward "/", PhoenixSwagger.Plug.SwaggerUI, swagger_url: "swagger.json"
  end
```

mbuhot commented:
I'm wondering if there is another way to opt-out of validation that doesn't involve matching multiple regexes against each request.

Eg, if we set a flag like `conn.private.phoenix_swagger.valid` in an earlier plug, then the validate plug can just pattern match on that flag and skip the validation.

soundmonster commented:
It seems to me that this plug has two interesting parts: validation of the request parameters against a Swagger Spec from ETS, and acting upon the results of the validation. Maybe it is a better idea to just extract `validate/1` and its dependencies from the plug into a standalone module, and provide a simple plug with its functionality as is in master right now. Then everyone can use the plug for simple scenarios, or build their own solution using `validate/1` in cases where custom error formatting is needed, etc.

soundmonster commented:
@0xAX @mbuhot thank you for your feedback! I have extracted request validation from the plug, so now it can be reused in custom plugs or called directly from controllers. Also, there is now a flag `conn.private.phoenix_swagger.valid` that, if true, would bypass the validator plug. I have added some info on how to use `Plug.Validate` and `ConnValidator` to the readme.

soundmonster commented:
@0xAX @mbuhot hi guys, I've rebased the PR on the latest master and made sure the tests pass. Any chance of getting this merged? The other open PR (#160) modifies the code that I have moved to ConnValidator - I'd be glad to help moving @pablobcb's changes to the new module.

mbuhot commented:
@soundmonster sorry this has been unmerged for so long. I'll take a look and try to merge this week. 👍

soundmonster commented:
Great! BTW, we're running off of my fork in production since late Oct 2017 and no problems so far 😉

mbuhot commented:
👍 Moving the query/path/header param validation out of the `Plug` and into a regular module looks good.

It might be a bit confusing for new users now that there are 3 ways to perform validation with `phoenix_swagger`.  I've created a separate isse #167 to refactor the README into guides, where we can make it clear when to use the `Plug.Validate` vs `Validator` vs `ConnValidate`
